From 2c4d93ce58d1a2bb90d98eb2476dcbf58dec8972 Mon Sep 17 00:00:00 2001
From: Andreas Baierl <ichgeh@imkreisrum.de>
Date: Thu, 28 Mar 2024 14:11:14 +0100
Subject: [PATCH] media: meson: vdec: reintroduce wiggle room

Without the wiggle room, it happens that matching offsets can't be found.
This results in non-matches and afterwards in frame drops in userspace apps.
Reintroduce this wiggle room again.

Tested on Odroid N2+ (S922X/G12B)

Fixes: 876f123b

Signed-off-by: Andreas Baierl <ichgeh@imkreisrum.de>
---
 drivers/staging/media/meson/vdec/vdec_helpers.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/media/meson/vdec/vdec_helpers.c b/drivers/staging/media/meson/vdec/vdec_helpers.c
index 7d2a75653250..62ef93690be3 100644
--- a/drivers/staging/media/meson/vdec/vdec_helpers.c
+++ b/drivers/staging/media/meson/vdec/vdec_helpers.c
@@ -370,7 +370,16 @@ void amvdec_dst_buf_done_offset(struct amvdec_session *sess,

 	/* Look for our vififo offset to get the corresponding timestamp. */
 	list_for_each_entry_safe(tmp, n, &sess->timestamps, list) {
-		if (tmp->offset > offset) {
+		s64 delta = (s64)offset - tmp->offset;
+
+		/* Offsets reported by codecs usually differ slightly,
+		 * so we need some wiggle room.
+		 * 4KiB being the minimum packet size, there is no risk here.
+		 */
+		if (delta > (-1 * (s32)SZ_4K) && delta < SZ_4K) {
+			match = tmp;
+			break;
+		} else {
 			/*
 			 * Delete any record that remained unused for 32 match
 			 * checks
@@ -379,10 +388,7 @@ void amvdec_dst_buf_done_offset(struct amvdec_session *sess,
 				list_del(&tmp->list);
 				kfree(tmp);
 			}
-			break;
 		}
-
-		match = tmp;
 	}

 	if (!match) {
--
2.30.2
